oauth2_http/java/com/google/auth/mtls/SecureConnectProvider.java

```diff
@@ -0,0 +1,161 @@
+ * library, with two notable differences: 1) All logic associated with parsing environment variables
+ * related to "mTLS usage" are omitted - a separate helper class will be introduced in the Gax
+ * library to serve this purpose. 2) getKeyStore throws {@link
+ * com.google.auth.mtls.CertificateSourceUnavailableException} instead of returning "null" if this
+ * cert source is not available on the device.
+ *
+ * <p>Additionally, this implementation will replace the existing "MtlsProvider" in the Gax library.
+ * The Gax library version of MtlsProvider will be marked as deprecated.
+ */
+public class SecureConnectProvider implements MtlsProvider {
+  interface ProcessProvider {
+    public Process createProcess(InputStream metadata) throws IOException;
+  }
+
+  static class DefaultProcessProvider implements ProcessProvider {
+    @Override
+    public Process createProcess(InputStream metadata) throws IOException {
+      if (metadata == null) {
+        return null;
+      }
+      List<String> command = extractCertificateProviderCommand(metadata);
+      return new ProcessBuilder(command).start();
+    }
+  }
+
+  private static final String DEFAULT_CONTEXT_AWARE_METADATA_PATH =
+      System.getProperty("user.home") + "/.secureConnect/context_aware_metadata.json";
+
+  private String metadataPath;
+  private ProcessProvider processProvider;
+
+  @VisibleForTesting
+  SecureConnectProvider(ProcessProvider processProvider, String metadataPath) {
+    this.processProvider = processProvider;
+    this.metadataPath = metadataPath;
+  }
+
+  public SecureConnectProvider() {
+    this(new DefaultProcessProvider(), DEFAULT_CONTEXT_AWARE_METADATA_PATH);
+  }
+
+  /** The mutual TLS key store created with the default client certificate on device. */
+  @Override
+  public KeyStore getKeyStore() throws IOException {
+    try (InputStream stream = new FileInputStream(metadataPath)) {
+      return getKeyStore(stream, processProvider);
+    } catch (InterruptedException e) {
+      throw new IOException("Interrupted executing certificate provider command", e);
+    } catch (GeneralSecurityException e) {
+      throw new CertificateSourceUnavailableException(
+          "SecureConnect encountered GeneralSecurityException:", e);
+    } catch (FileNotFoundException exception) {
+      // If the metadata file doesn't exist, then there is no key store, so we will throw sentinel
+      // error
+      throw new CertificateSourceUnavailableException("SecureConnect metadata does not exist.");
+    }
+  }
+
+  @VisibleForTesting
+  static KeyStore getKeyStore(InputStream metadata, ProcessProvider processProvider)
+      throws IOException, InterruptedException, GeneralSecurityException {
+    Process process = processProvider.createProcess(metadata);
+
+    // Run the command and timeout after 1000 milliseconds.
+    int exitCode = runCertificateProviderCommand(process, 1000);
```

Comment:
I know this is from existing code in Gax, but do you know any reasoning to this 1000 number that we can add to comment?

Reply:
Unfortunately not, it seems like an arbitrary interval based on the fact that the command usually finishes instantly or hangs. I guess I can mention this in the comment.